### PR TITLE
Experiment b commonconst

### PR DESCRIPTION
**Pull request by iritkatriel**
TODO:

1. ~~Add the tuples~~
2. ~~PREDICT stuff in ceval.c~~
3. ~~fix test_modulefinder~~
4. ~~replace LOAD_ASSERTION_ERROR~~
5. Review the tests to see if any are now missing

### CODE REVIEW

**Review by gvanrossum — COMMENTED** [review]
Oh, that looks cool! Maybe the prediction machinery should be directed here too?

**Comment by iritkatriel** [comment]
> the prediction machinery

The who?

**Comment by gvanrossum** [comment]
Macros like PREDICT() and PREDICTED().

**Comment by iritkatriel** [comment]
ah yes, things like this:
    ```
    ADDOP(c, GET_YIELD_FROM_ITER);
        ADDOP_LOAD_CONST(c, Py_None);
```

and then `GET_YIELD_FROM_ITER` predicts LOAD_CONST..

**Comment by gvanrossum** [comment]
So some of those will need to predict LOAD_COMMON_CONST instead. Maybe the
more general LOAD_CONST no longer occurs often enough to require predicting?

**Comment by iritkatriel** [comment]
Yes, I'll check each case. In the case I quoted the constant is None so it's a common const for sure.

**Comment by iritkatriel** [comment]
I have some issues with when/where I allocated the common consts data - refleaks tests show I allocated it more than once (and if I try not to then subinterpreters don't have it). I'll have to rethink where this actually sits (I put it next to the small_ints but allocate in another place because I need the AttributeError exception etc to be ready).

I need to go soon so will continue tomorrow.

File: Python/common_consts.c [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        return -1;
+    int ret = add_common_const(index, obj);
+    Py_DECREF(obj);
+    return ret;
+}
+
+static int
+add_common_string(int index, const char *s)
+{
+    PyObject *obj = PyUnicode_FromString(s);
```
Comment (methane) [inline Python/common_consts.c:86]:
```suggestion
    PyObject *obj = PyUnicode_InternFromString(s);
```

**Comment by gvanrossum** [comment]
Did you make any progress on the leak? Where you put the initialization looks fine to me (but I'm no expert on that part of the interpreter).

**Comment by iritkatriel** [comment]
I'm just looking at the leak. I'm moving the dict to the interpreter state alongside the array, so that they have the same lifetime. And I'm putting the alloc.dealloc alongside PyLong_Init/Fini. (I'm aiming to make this alloc/dealloc like small_ints).

That said, I'm not sure why the small_ints array is allocated on the interpreter state. Wouldn't one systemwide copy be enough?

**Comment by markshannon** [comment]
What's the justification for adding the seemingly arbitrary selection of tuples and strings rather than just a few simple constants?
Could we just start with the simple constants, and add the others later?

Why the procedural approach to building the table of constants, instead of declaring `PyObject *common_consts[] = { ...`?

**Comment by iritkatriel** [comment]
> What's the justification for adding the seemingly arbitrary selection of tuples and strings rather than just a few simple constants?

This is the list Guido extracted of the common constants in PyPl:
https://github.com/faster-cpython/ideas/issues/65#issuecomment-878637853 

> Why the procedural approach to building the table of constants, instead of declaring PyObject *common_consts[] = { ...?

Error checking.

**Comment by markshannon** [comment]
> 
> > Why the procedural approach to building the table of constants, instead of declaring PyObject *common_consts[] = { ...?
> 
> Error checking.

I don't understand. There can be no runtime errors in a declaration, only in executable code.

**Comment by iritkatriel** [comment]
> > > Why the procedural approach to building the table of constants, instead of declaring PyObject *common_consts[] = { ...?
> > 
> > 
> > Error checking.
> 
> I don't understand. There can be no runtime errors in a declaration, only in executable code.

Are PyLong_FromLong() and PyFloat_FromFloat() not executable code?

**Comment by markshannon** [comment]
> Are PyLong_FromLong() and PyFloat_FromFloat() not executable code?

Yes, they are. So don't use them :slightly_smiling_face: 

`PyObject *common_consts[] = { Py_None, Py_Ellipsis, PyExc_AssertionError, ... };`

You could add a few float constants, `PyFloat_Zero`, `PyFloat_One`.
Integers are probably best handled with their own instruction, e.g. `LOAD_SMALL_INT`.

**Comment by gvanrossum** [comment]
> `PyObject *common_consts[] = { Py_None, Py_Ellipsis, PyExc_AssertionError, ... };`

Note that, as a static declaration, the Windows C compiler doesn’t accept that. (C++ does.)

**Comment by markshannon** [comment]
> > `PyObject *common_consts[] = { Py_None, Py_Ellipsis, PyExc_AssertionError, ... };`
> 
> Note that, as a static declaration, the Windows C compiler doesn’t accept that. (C++ does.)

Are you sure about that? 
How does [this](https://github.com/python/cpython/blob/main/Objects/object.c#L1666) compile?

**Comment by gvanrossum** [comment]
>  How does [this](https://github.com/python/cpython/blob/main/Objects/object.c#L1666) compile?

Hm, it does compile. I am sure I saw a problem with references to type objects in C files (that I made go away by compiling as C++), but I must have misunderstood its root cause.

Or maybe we compile with some flag that made the error go away. I have a simple example where I initialize a struct member with `&_Py_NoneStruct` (i.e. the expansion of `Py_None`), and VS Code highlights the `&` character and tells me "expression must have a constant value C/C++(28)", but it compiles. So now I'm officially confused.

Possibly the root of my confusion is the common idiom of initializing type objects in *extension modules* using `PyVarObject_HEAD_INIT(NULL, 0)` or `PyObject_HEAD_INIT(NULL)`, and then setting the actual type using `PyType_Ready()`. I somehow thought that was due to MSVC, but it may be due to the DLL linker or possibly needed for shared libraries in general.

Anyway, just ignore me, that's always been the best option.

**Pull request closed without merging**